core/src/dird/ua_restore.cc

```diff
@@ -839,15 +845,67 @@ static bool get_date(UaContext* ua, char* date, int date_len)
   ua->SendMsg(
       _("The restored files will the most current backup\n"
         "BEFORE the date you specify below.\n\n"));
+  std::string cmpdate;
   for (;;) {
     if (!GetCmd(ua, _("Enter date as YYYY-MM-DD HH:MM:SS :"))) { return false; }
-    if (StrToUtime(ua->cmd) != 0) { break; }
+    cmpdate = CompensateShortDate(ua->cmd);
+
+    if (StrToUtime(cmpdate.c_str()) != 0) { break; }
     ua->ErrorMsg(_("Improper date format.\n"));
   }
-  bstrncpy(date, ua->cmd, date_len);
+  bstrncpy(date, cmpdate.c_str(), date_len);
   return true;
 }
 
+/**
+ * Compensate missing date-time elements in shorter formats
+ * Returns a full compensated date when argument has a correct short format
+ * Returns the argument unchanged, if there is a format problem
+ */
+
+std::string CompensateShortDate(const char* cmd)
+{
+  struct tm tm;
+  char* garbage = nullptr;
+
+  memset(&tm, -1, sizeof(tm));
+  if ((sscanf(cmd, "%u-%u-%u %u:%u:%u%s", &tm.tm_year, &tm.tm_mon, &tm.tm_mday,
+              &tm.tm_hour, &tm.tm_min, &tm.tm_sec, &garbage)
+           == 7
+       || sscanf(cmd, "%u-%u-%u %u:%u%s", &tm.tm_year, &tm.tm_mon, &tm.tm_mday,
+                 &tm.tm_hour, &tm.tm_min, &garbage)
+              == 6
+       || sscanf(cmd, "%u-%u-%u %u%s", &tm.tm_year, &tm.tm_mon, &tm.tm_mday,
+                 &tm.tm_hour, &garbage)
+              == 5
+       || sscanf(cmd, "%u-%u-%u%s", &tm.tm_year, &tm.tm_mon, &tm.tm_mday,
+                 &garbage)
+              == 4
+       || sscanf(cmd, "%u-%u%s", &tm.tm_year, &tm.tm_mon, &garbage) == 3
+       || sscanf(cmd, "%u%s", &tm.tm_year, &garbage) == 2)
+      && (garbage == nullptr)) {
+    if (tm.tm_mon == -1) { tm.tm_mon = 1; }
+    if (tm.tm_mday == -1 || tm.tm_mday == 0) { tm.tm_mday = 1; }
+    if (tm.tm_hour == -1) { tm.tm_hour = 0; }
+    if (tm.tm_min == -1) { tm.tm_min = 0; }
+    if (tm.tm_sec == -1) { tm.tm_sec = 0; }
```

Comment:
if you zero-initialized `tm`, then this would simplify to:
```suggestion
    if (tm.tm_mon == 0) { tm.tm_mon = 1; }
    if (tm.tm_mday == 0) { tm.tm_mday = 1; }
```